calico_kubernetes/policy.py

```diff
@@ -0,0 +1,343 @@
+# Copyright 2015 Metaswitch Networks
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import re
+import logging
+from netaddr import IPNetwork, AddrFormatError
+from pycalico.datastore_datatypes import Rule
+
+_log = logging.getLogger(__name__)
+
+
+class PolicyParser(object):
+    """
+    Class which parses Kubernetes annotation-based policy strings and
+    converts them to libcalico Rules objects.
+
+    Usage:
+      parser = PolicyParser(namespace)
+      rule = parser.parse_line("allow tcp from label stage=prod")
+    """
+    def __init__(self, namespace):
+        self.namespace = namespace
+
+    def parse_line(self, policy):
+        """
+        Takes a single line of policy as defined in the annotations of a
+        pod and returns the equivalent libcalico Rule object.
+        :param policy: Policy string from annotations.
+        :return: A Rule object which represent the given policy.
+        """
+        _log.info("Parsing policy line: '%s'", policy)
+        splits = self.validate_policy(policy)
+    
+        # Parse based on the protocol.
+        if "tcp" in splits:
```

Comment:
I hate to drop this feedback on a fully-formed PR... but I am not sure about this approach. I think we should be using a proper parsing library to build up an AST for our rules.

The cheapo implementation that I was imagining was to use docopt to build this (we do have a working docopt parser for the command already, it's in the calicoctl Usage string). That way we guarantee that the command will be parsed in the same way as calicoctl would (assuming that the docstrings don't diverge).

The less cheap way would be to use [Pyparsing](https://pyparsing.wikispaces.com/), which lets you build arbitrary parsers in a clean pythonic way. Then you can do something like

```python
parser = Optional(protocol) + Optional(from_clause) + Optional(to_clause)
```

Your approach clearly works for the test cases that you've written, and it may well be correct, but it's much less obviously correct than the parser-based approaches above, which would probably be tens of lines of code on our side.

Thoughts?